contracts/infrastructure/DappRegistry.sol

```diff
@@ -2,77 +2,212 @@ pragma solidity ^0.6.12;
+        delete authorisations[_registryId][_dapp];
+        emit DappRemoved(_registryId, _dapp);
+    }
+
+    /**
+    * @notice Request to change an authorisation filter for a dapp that has previously been authorised. We cannot 
+    * immediately override the existing filter and need to store the new filter for a timelock period before being 
+    * able to change the filter.
+    * @param _registryId The id of the registry to modify
+    * @param _dapp The address of the dapp contract to authorise.
+    * @param _filter The address of the new filter contract to use.
+    */
+    function requestFilterUpdate(uint8 _registryId, address _dapp, address _filter) external onlyOwner(_registryId) {
+        require(authorisations[_registryId][_dapp] != bytes32(0), "AR: unknown dapp");
+        uint validAfter = block.timestamp + timelockPeriod;
+        // Store the future authorisation as {filter:160}{validAfter:64}
+        pendingFilterUpdates[_registryId][_dapp] = bytes32((uint(uint160(_filter)) << 64) | validAfter);
+        emit FilterUpdateRequested(_registryId, _dapp, _filter, validAfter);
+    }
+
+    /**
+    * @notice Confirm the filter change requested by `requestFilterUpdate`
+    * @param _registryId The id of the registry to modify
+    * @param _dapp The address of the dapp contract to authorise.
+    */
+    function confirmFilterUpdate(uint8 _registryId, address _dapp) external {
+        uint newAuth = uint(pendingFilterUpdates[_registryId][_dapp]);
+        require(newAuth > 0, "AR: no pending filter update");
+        uint validAfter = newAuth & 0xffffffffffffffff;
+        require(validAfter <= block.timestamp, "AR: too early to confirm auth");
+        authorisations[_registryId][_dapp] = bytes32(newAuth);
+        emit FilterUpdated(_registryId, _dapp, address(uint160(newAuth >> 64)), validAfter); 
+        delete pendingFilterUpdates[_registryId][_dapp];
+    }
+
+    /********  Internal Functions ***********/
+
+    function validateOwner(uint8 _registryId) internal view {
+        address owner = registryOwners[_registryId];
+        require(owner != address(0), "AR: unknown registry");
+        require(msg.sender == owner, "AR: sender != registry owner");
     }
 
-    function getFilter(address _wallet, address _contract) internal view returns (bool, address) {
-        // check argent default registry first (enbaled by default, i.e. false => true)
-        if (!isEnabled[_wallet][ARGENT_DEFAULT] && authorisations[ARGENT_DEFAULT][_contract].isActive) {
-            return (true, authorisations[ARGENT_DEFAULT][_contract].filter);
-        } else {
-            for (uint i = 0; i < registries.length; i++) {
-                if (isEnabled[_wallet][registries[i]] && authorisations[registries[i]][_contract].isActive) {
-                    return (true, authorisations[registries[i]][_contract].filter);
+    function getFilter(address _wallet, address _dapp) internal view returns (bool, address) {
```

Comment:
A valid point raised by the client teams is that if a dapp is authorised on multiple registries but with different filters (e.g. registry[0] authorises buy/sell on Aave, while registry[1] athorises staking) the second one will be ignored.

We could fix that by having the `getFilter` method return an array of matching filters, and loop through the filters in `isAuthorised` until one of them passes.

Reply:
I don't think it's worth the added gas cost of having to systematically do N reads for N filters. Besides, philosophically, I think of a filter as the answer to the question "what is the set of method calls on this contract such that all assets stay under the control of the wallet?". There should be a precise answer to this question for each dapp. So in the case of Aave, after reviewing the Aave contract, we may determine that minting, redeeming _and_ staking are all permissible actions, and would define a filter that reflects that. Partial filters split across registries would in my opinion indicate that multiple registries disagree on the above question. But such disagreement should be resolved off-chain IMO.